File: cocos2dx/CocosLuaGame/src/app/classes/game_character.lua

local ACTION_IDLE = "Walk"
local ACTION_WALK = "Walk"
local ACTION_RUN = "Walk"
local ACTION_ATTACK = "Fire"

--local AnimationController = require("app.components.animation_controller")
local StateMachine = require("app.components.state_machine")

local OBJECT_NAME = "GameCharacter"

local ObjectClass = class(OBJECT_NAME, function() 
    local obj = GameEntity:create() 
    return obj
end)

ObjectClass.DIRECTION_LEFT = -1
ObjectClass.DIRECTION_RIGHT = 1

function ObjectClass:ctor(animationName)
    -- animation controller
    local ani = AnimationController:create()
    self:addComponent(ani)
    self.ani = ani
    
    self.ani:load(animationName)
    self.ani:setMovementEventCallFunc(function(movementType, movementID) 
        if movementID == ACTION_ATTACK then
            if movementType == ccs.MovementEventType.complete or
                movementType == ccs.MovementEventType.loopComplete then
                self.fsm:doEvent("doIdle")
            end
        end
    end)
    
    -- state machine
    local cfg = {
        events =
        {
            { name="doStartup", from=StateMachine.STATE_READY, to="idle" },
            { name="doIdle", from=nil, to="idle" },
            { name="doWalk", from="idle", to="walk" },
            { name="doWalk", from="run", to="walk" },
            { name="doRun", from="idle", to="run" },
            { name="doRun", from="walk", to="run" },
            { name="doAction1", from={"idle", "run", "walk"}, to="action1" },
        },

        callbacks =
        {
            on_idle = function(event) self:onIdle() end,
            on_walk = function(event) self:onWalk() end,
            on_run = function(event) self:onRun() end,
            on_action1 = function(event) self:onAttack() end,
        },

    }
    local fsm = StateMachine.new()
    fsm:setupState(cfg)
    self:addComponent(fsm)
    self.fsm = fsm
    
    -- position
    self.desiredPositionX = self:getPositionX()
    self.desiredPositionY = self:getPositionY()
    
    -- node event
    self:registerScriptHandler(function(event)
        if event == "enter" then
            self:onEnter()
        end
    end)
    

    
    -- schedule update
    self:scheduleUpdateWithPriorityLua(handler(self, self.onFrame), 0)
end

function ObjectClass:getContentSize()
    return self.ani:getArmature():getContentSize()
end

function ObjectClass:onEnter()
    self.fsm:doEvent("doStartup")
end

function ObjectClass:onFrame(dt)
    --[[
    local boneDict = self.ani.armature:getBoneDic()
    for type, bone in pairs(boneDict) do
        printInfo("%s = %s", tostring(type), tostring(bone))
        local detector = bone:getColliderDetector()
        if detector then
            local bodyList = ccs.ColliderDetector:getColliderBodyList
        end
    end
    --]]
    --self.ani.armature:drawContour()

    
    local state = self.fsm:getState()
    
    if state == "idle" then
        --self:fsm:doEvent("doWalk")
    elseif state == "walk" then
        local speed = 2
        local direction = self:getDirection()      
        self.desiredPositionX = self:getPositionX() + speed * direction
    elseif state == "run" then
        local speed = 4
        local direction = self:getDirection()      
        self.desiredPositionX = self:getPositionX() + speed * direction
    end
end

function ObjectClass:onIdle()
    self.ani:play(ACTION_IDLE)
end

function ObjectClass:onWalk()
    self.ani:play(ACTION_WALK)
end

function ObjectClass:onRun()
    self.ani:play(ACTION_RUN)
end

function ObjectClass:onAttack()
    self.ani:play(ACTION_ATTACK)
end

function ObjectClass:doIdle()
    if self.fsm:canDoEvent("doIdle") then self.fsm:doEvent("doIdle") end
end

function ObjectClass:doWalk(direction)
    if self.fsm:canDoEvent("doWalk") then 
        self:changeDirection(direction)
        self.fsm:doEvent("doWalk") 
    end
end

function ObjectClass:doRun(direction)
    if self.fsm:canDoEvent("doRun") then 
        self:changeDirection(direction)
        self.fsm:doEvent("doRun") 
    end
end

function ObjectClass:doAttack()
    if self.fsm:canDoEvent("doAction1") then 
        self.fsm:doEvent("doAction1") 
    end
end

function ObjectClass:getDirection()
    -- TODO: 该资源方向刚好相反
    --if self:getScaleX() * ObjectClass.DIRECTION_RIGHT < 0 then
    if self:getScaleX() * ObjectClass.DIRECTION_RIGHT > 0 then
        return ObjectClass.DIRECTION_LEFT
    else
        return ObjectClass.DIRECTION_RIGHT
    end

    --return self:getScaleX()>0 and ObjectClass.DIRECTION_RIGHT or ObjectClass.DIRECTION_LEFT
end

function ObjectClass:changeDirection(direction)
    local current_direction = self:getDirection()
    if current_direction * direction < 0 then
        self:setScaleX(self:getScaleX() * (-1))
    end
end

function ObjectClass:getDesiredPosition()
    return self.desiredPositionX, self.desiredPositionY
end

function ObjectClass:setDesiredPosition(x, y)
    self.desiredPositionX = x
    self.desiredPositionY = y
end

function ObjectClass:checkCollider(rect)
    return self.ani:checkCollider(rect)
end

return ObjectClass

File: cocos2dx/top/Classes/GameCharacter.cpp
#include "GameCharacter.h"

#include "AnimationController.h"

USING_NS_CC;

GameCharacter * GameCharacter::create( const char *animationName )
{
	auto pRet = new GameCharacter();
	if (pRet && pRet->init(animationName)) 
	{ 
		pRet->autorelease(); 
		return pRet; 
	} 
	else 
	{ 
		delete pRet; 
		pRet = NULL; 
		return NULL;
	} 
}

GameCharacter::GameCharacter()
	: m_direction(Direction::Right)
{

}

GameCharacter::~GameCharacter()
{

}

bool GameCharacter::init( const char *animationName )
{
	if (!Node::init())
	{
		return false;
	}

	auto animationController = AnimationController::create();
	this->addComponent(animationController);

	animationController->load(animationName);
	animationController->setMovementCallback([=](cocostudio::Armature *armature, cocostudio::MovementEventType movementType, const std::string& movementID) {
		if (movementID.compare("attack_down") == 0)
		{
			if (movementType == cocostudio::MovementEventType::LOOP_COMPLETE || movementType == cocostudio::MovementEventType::COMPLETE)
			{
				this->doIdle();
			}
		}
	});

	auto fsm = StateMachine::create();
	this->addComponent(fsm);

	// init finite state machine
	fsm->setupState("doStartup", StateMachine::STATE_NONE, State::Idle);	
	fsm->setupState("doIdle", State::Walk, State::Idle);
	fsm->setupState("doIdle", State::Run, State::Idle);
	fsm->setupState("doWalk", State::Idle, State::Walk);
	fsm->setupState("doWalk", State::Run, State::Walk);
	fsm->setupState("doRun", State::Idle, State::Run);
	fsm->setupState("doRun", State::Walk, State::Run);
	fsm->setupState("doAttack", StateMachine::STATE_ALL, State::Attack);

	fsm->registerEnterStateCallback(State::Idle, [this](const char *eventName, int fromState, int toState) {
		this->onIdle();
		return true;
	});
	fsm->registerEnterStateCallback(State::Walk, [this](const char *eventName, int fromState, int toState) {
		this->onWalk();
		return true;
	});
	fsm->registerEnterStateCallback(State::Run, [this](const char *eventName, int fromState, int toState) {
		this->onRun();
		return true;
	});
	fsm->registerEnterStateCallback(State::Attack, [this](const char *eventName, int fromState, int toState) {
		this->onAttack();
		return true;
	});

	this->scheduleUpdate();

	return true;
}

void GameCharacter::update( float dt )
{
	Node::update(dt);
}

AnimationController * GameCharacter::getAnimationController()
{
	return dynamic_cast<AnimationController *>(this->getComponent(AnimationController::NAME));
}

StateMachine * GameCharacter::getStateMachine()
{
	return dynamic_cast<StateMachine *>(this->getComponent(StateMachine::NAME));
}


GameCharacter::Direction GameCharacter::getDirection()
{
	return m_direction;	
}

void GameCharacter::changeDirection( Direction direction )
{
	if (m_direction != direction)
	{
		this->setScaleX(this->getScaleX() * (-1));
		m_direction = direction;
	}
}

void GameCharacter::doIdle()
{
	if (this->getStateMachine()->getState() != State::Idle)
	{
		this->getStateMachine()->doEvent("doIdle");
	}
}

void GameCharacter::doWalk()
{
	if (this->getStateMachine()->getState() != State::Walk)
	{
		this->getStateMachine()->doEvent("doWalk");
	}
}

void GameCharacter::doRun()
{
	if (this->getStateMachine()->getState() != State::Run)
	{
		this->getStateMachine()->doEvent("doRun");
	}
}

void GameCharacter::doAttack()
{
	if (this->getStateMachine()->getState() != State::Attack)
	{
		this->getStateMachine()->doEvent("doAttack");
	}
}

void GameCharacter::onIdle()
{
	log("on idle");
	this->getAnimationController()->play("idle");
}

void GameCharacter::onWalk()
{
	log("on walk");
	this->getAnimationController()->play("walk");
}

void GameCharacter::onRun()
{
	log("on run");
	this->getAnimationController()->play("run");
}

void GameCharacter::onAttack()
{
	log("on attack");
	this->getAnimationController()->play("attack_down");
}

